Bootstrap repos

alexdutton opened the pull request:
This adds a Python 3 script (alongside `add-repos-from-github.py`) that bootstraps a proto-commons repo and then creates a remote repo on GitHub and pushes to it, with topics set appropriately. Closes #2.

alexdutton committed:
Script to bootstrap a new repo based on a country code and push it to GitHub

mhl commented on create-repo.py:
```diff
@@ -0,0 +1,93 @@
+
+print(query)
+
+response = requests .post(sparql_endpoint, query,
```
Tiny thing, obviously but there's an extraneous space here (I'd have hoped this kind of thing would be picked up by live linting in your editor?)

alexdutton replied:
Was using vanilla Atom at the time, which in retrospect wasn't too helpful for Python. Fixed this and other style issues with the PEP8 linting in PyCharm :-)

mhl commented on create-repo.py:
```diff
@@ -0,0 +1,93 @@
+if 'en' not in languages:
+    languages.append('en')
+
+country_wikidata_id = bindings[0]['country']['value'].split('/')[-1]
```
It might be worth checking that there's exactly one `?country` found by the query (and erroring otherwise) before picking the country from the first result row. (I haven't checked, but I bet that at some point there'll be multiple items given the same `P279` value...)

alexdutton replied:
I've added explicit checks for 0 and >1 countries, with appropriate messages.

mhl commented on create-repo.py:
```diff
@@ -0,0 +1,93 @@
+# * strip any leading or trailing hyphens
+repo_name = 'proto-commons-' + re.sub('[^a-z]+', '-', label.lower().replace("'", '')).strip('-')
+
+print(languages)
```
Generally, with bare `print` calls like this, which are really here to help us check that nothing's going wrong, I'd always prefix them with an indication of what's expected, so if it's unexpectedly empty, you can see what was missing. e.g. `print('languages:', languages)`, etc.

alexdutton replied:
This whole section will now produce output something like:

```
Country information:
  Languages:       el tr en
  Country ID:      Q229
  Country name:    Cyprus
  Repository name: proto-commons-cyprus
```

mhl commented on create-repo.py:
```diff
@@ -0,0 +1,93 @@
+    }, f, indent=2)
+
+for dirname in ['boundaries/build', 'executive', 'legislative']:
+    dirname = os.path.join(iso_3166_1_code, dirname)
```
To make this slightly easier to read, I'd use a new variable name here rather than rebinding `dirname`.

mhl commented on create-repo.py:
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
```
Since this has a useful shebang line, it'd be good to mark it as executable as well.

alexdutton replied:
Done!

mhl reviewed (dismissed):
I've requested a few small changes, but otherwise things look good to me :+1: The f-strings are cool, but I'd probably change them to normal `.format`s since I can imagine we're going to want to run this in situations where such a recent Python 3 version is going to be awkward to arrange (e.g. on our servers)

alexdutton commented:
I've addressed all @mhl's feedback, and also:

* used `argparse` for argument parsing, and for a helpful `--help`
* `s/iso_3166_1_code/repo_dir/` when the ISO code is being used as a directory, and added the ability to explicitly specify the directory name (which will be required if/when we support taking a Wikidata ID instead of a country code in future)
* Some more helpful print statements, including a link to the created repository at the end
* Removed the f-strings in favour of `.format(…)` to support Py3 versions earlier than Py3.6.

The changes since Mark's review are in a fixup commit, and I'll squash all three commits together before merging; in retrospect I don't think there's much point in separating out the GitHub manipulation into a separate commit when it's clear that it's an integral part of the required functionality.

The base will also be changed to `master` before merging, which will be after #5 is merged.

mhl reviewed (approved):
Looks good! :sparkles: :+1:
